source/postprocess/mobility_statistics.cc

```diff
@@ -0,0 +1,137 @@
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/simulator.h>
+#include <aspect/postprocess/mobility_statistics.h>
+#include <aspect/material_model/simple.h>
+#include <aspect/global.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+#include <aspect/geometry_model/spherical_shell.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    template <int dim>
+    std::pair<std::string,std::string>
+    MobilityStatistics<dim>::execute (TableHandler &statistics)
+    {
+      const QGauss<dim> quadrature_formula (this->get_fe()
+                                            .base_element(this->introspection().base_elements.velocities).degree+1);
+      const unsigned int n_q_points = quadrature_formula.size();
+
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature_formula,
+                               update_values   |
+                               update_quadrature_points |
+                               update_JxW_values);
+      std::vector<Tensor<1,dim>> velocity_values(n_q_points);
+
+      double local_velocity_square_integral = 0;
+      double local_velocity_square_integral_top_boundary;
+      double local_top_boundary_area;
+      const std::set<types::boundary_id>
+      boundary_indicators
+        = this->get_geometry_model().get_used_boundary_indicators ();
+      const types::boundary_id top_boundary_id = this->get_geometry_model().translate_symbolic_boundary_name_to_id("top");
+
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned())
+          {
+            fe_values.reinit (cell);
+            fe_values[this->introspection().extractors.velocities].get_function_values (this->get_solution(),
+                                                                                        velocity_values);
+            for (unsigned int q = 0; q < n_q_points; ++q)
+              {
+                local_velocity_square_integral += ((velocity_values[q] * velocity_values[q]) *
+                                                   fe_values.JxW(q));
+              }
+            //now for the surface cells only
+            for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+              if (cell->face(f)->boundary_id() == top_boundary_id)
+                {
+                  // determine the squared velocity on the face
+                  double local_sqvel = 0.0;
+                  double local_fe_face_area = 0.0;
+                  for (unsigned int q = 0; q < n_q_points; ++q)
+                    {
+                      const double vel_mag = velocity_values[q].norm();
+                      local_sqvel += ((vel_mag * vel_mag) * fe_values.JxW(q));
+                      local_fe_face_area += fe_values.JxW(q);
```

Comment:
You can not use an `fe_values` object to compute properties on faces of cells. `fe_values` is meant to evaluate or integrate the solution over the volume of cells, while `fe_face_values` is meant to integrate over the surface of cells. Keep everything in the one loop you have now, but above you need to keep an `fe_face_values` object and the corresponding quadrature rule.

Reply:
Ah yes